compiler/circle-mpqsolver/src/utils/ModuleCloner.cpp

```diff
@@ -0,0 +1,172 @@
+  {
+    if (output_node->dim(r).known())
+      output_shape->dim(r).set(output_node->dim(r).value());
+  }
+  graph_output->shape(std::move(output_shape));
+}
+
+std::unique_ptr<loco::Graph> clone_graph(loco::Graph *graph_org)
+{
+  luci::CloneContext clonectx;
+
+  auto graph = loco::make_graph();
+  auto graph_clone = graph.get();
+  auto &graph_name = graph_org->name();
+
+  graph_clone->name(graph_name);
+
+  // clone inputs
+  auto inputs = graph_org->inputs();
+  assert(inputs);
+  for (uint32_t n = 0; n < inputs->size(); ++n)
+  {
+    auto input_org = luci::input_node(graph_org, n);
+    assert(input_org != nullptr);
+
+    auto *input_clone = graph_clone->nodes()->create<luci::CircleInput>();
+    luci::copy_common_attributes(input_org, input_clone);
+
+    add_graph_input(graph_clone, input_clone);
+    clonectx.emplace(input_org, input_clone);
+  }
+
+  // clone nodes
+  auto nodes = graph_org->nodes();
+  for (uint32_t n = 0; n < nodes->size(); ++n)
+  {
+    auto node = nodes->at(n);
+
+    // skip for CircleInput, CircleOutput
+    if (dynamic_cast<luci::CircleInput *>(node) != nullptr)
+      continue;
+    if (dynamic_cast<luci::CircleOutput *>(node) != nullptr)
+      continue;
+
+    auto node_org = loco::must_cast<luci::CircleNode *>(node);
+    assert(clonectx.find(node_org) == clonectx.end());
+
+    auto *node_clone = clone_node(node_org, graph_clone);
+    clonectx.emplace(node_org, node_clone);
+  }
+
+  // connect nodes
+  for (uint32_t n = 0; n < nodes->size(); ++n)
+  {
+    auto node = nodes->at(n);
+
+    // skip for CircleInput, CircleOutput
+    if (dynamic_cast<luci::CircleInput *>(node) != nullptr)
+      continue;
+    if (dynamic_cast<luci::CircleOutput *>(node) != nullptr)
+      continue;
+
+    auto node_org = loco::must_cast<luci::CircleNode *>(node);
+    clone_connect(node_org, clonectx);
+  }
+
+  // clone outputs
+  for (uint32_t n = 0; n < graph_org->outputs()->size(); ++n)
+  {
+    auto output_org = luci::output_node(graph_org, n);
+    assert(output_org != nullptr);
+
+    auto *output_clone = graph_clone->nodes()->create<luci::CircleOutput>();
+    luci::copy_common_attributes(output_org, output_clone);
+    // note: we don't add output_clone to clonectx.
+    // logically, output is not used as an input to any other nodes.
+    auto output_from = loco::must_cast<luci::CircleNode *>(output_org->from());
+    auto it = clonectx.find(output_from);
+    assert(it != clonectx.end());
+    output_clone->from(it->second);
+
+    add_graph_output(graph_clone, output_clone);
+  }
+
+  return graph;
+}
+
+} // namespace
+
+std::unique_ptr<luci::Module> ModuleCloner::Clone(const luci::Module *module)
+{
+  std::unique_ptr<luci::Module> new_module = luci::make_module();
+  for (size_t idx = 0; idx < module->size(); ++idx)
+  {
+    auto graph = module->graph(idx);
+    auto new_graph = clone_graph(graph);
+    if (!graph)
+    {
+      return nullptr;
+    }
```

Comment:
Checking first would reduce the costs.
```suggestion
    if (!graph)
    {
      return nullptr;
    }
    auto new_graph = clone_graph(graph);
    
```

Reply:
Well, if `clone_graph` returns `nullptr`, both way have almost same costs.